File: 백준/Gold/15686. 치킨 배달/치킨 배달.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

struct Point {
    int x, y;

    Point(int x, int y) : x(x), y(y) {}
};

int N, M, answer = 987654321;
vector<Point> chickens, houses, searches;


int visited[2501];

void bfs() {
    int chickenDist = 0;

    for (auto house: houses) {
        int sum = 987654321;
        for (auto search: searches) {
            sum = min(abs(house.x - search.x) + abs(house.y - search.y), sum);
        }
        chickenDist += sum;
    }
    answer = min(chickenDist, answer);
}

void dfs(int index, int cnt) {
    if (cnt == M) {
        bfs();
        return;
    }

    for (int i = index; i < chickens.size(); i++) {
        if (!visited[i]) {
            visited[i] = 1;
            searches.emplace_back(chickens[i]);
            dfs(i, cnt + 1);
            searches.pop_back();
            visited[i] = 0;
        }
    }


}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> N >> M;
    vector<vector<int>> map(51, vector<int>(51));
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= N; j++) {
            cin >> map[i][j];
            if (map[i][j] == 1) houses.emplace_back(i, j);
            if (map[i][j] == 2) chickens.emplace_back(i, j);
        }
    }
    dfs(0, 0);
    cout << answer;
}

File: 백준/Gold/15686. 치킨 배달/치킨 배달.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int[][] city;
    static ArrayList<Point> chickens;
    static ArrayList<Point> houses;
    static Point[] selectedChickens;
    static boolean[][] visited;
    static int answer = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer tokenizer = new StringTokenizer(br.readLine());
        N = Integer.parseInt(tokenizer.nextToken());
        M = Integer.parseInt(tokenizer.nextToken());
        city = new int[N][N];
        visited = new boolean[N][N];
        chickens = new ArrayList<>();
        houses = new ArrayList<>();
        selectedChickens = new Point[M];
        for (int i = 0; i < N; i++) {
            tokenizer = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                city[i][j] = Integer.parseInt(tokenizer.nextToken());
                if (city[i][j] == 1) {
                    houses.add(new Point(i, j));
                } else if (city[i][j] == 2) {
                    chickens.add(new Point(i, j));
                }
            }
        }

        findChickenStore(0, 0);
        bw.write(answer + "");
        bw.flush();
        bw.close();
        br.close();
    }

    private static void findChickenStore(int index, int depth) {
        if (depth == M) {
            answer = Math.min(getChickenDistance(), answer);
            return;
        }

        for (int i = index; i < chickens.size(); i++) {
            int x = chickens.get(i).x;
            int y = chickens.get(i).y;
            if (!visited[x][y]) {
                visited[x][y] = true;
                selectedChickens[depth] = new Point(x, y);
                findChickenStore(i + 1, depth + 1);
                selectedChickens[depth] = null;
                visited[x][y] = false;
            }
        }
    }

    private static int getChickenDistance() {
        int result = 0;
        for (Point house : houses) {
            int sum = Integer.MAX_VALUE;
            for (Point selectedChicken : selectedChickens) {
                int dist = Math.abs(house.x - selectedChicken.x) + Math.abs(house.y - selectedChicken.y);
                sum = Math.min(dist, sum);
            }
            result += sum;
        }
        return result;
    }

    static class Point {
        int x, y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }
}